crates/holochain_cascade/src/lib.rs

```diff
@@ -774,49 +748,65 @@ impl CascadeImpl {
-        let authority = self.am_i_an_authority(action_hash.clone().into()).await?;
-        if !(authoring || authority) {
-            self.fetch_record(action_hash.into(), options.into())
-                .await?;
-        }
-
-        // Check if we have the data now after the network call.
-        self.cascading(query).await
+        self.get_with_query(query, action_hash.into(), options)
+            .await
     }
 
-    #[cfg_attr(feature = "instrument", tracing::instrument(skip(self, options)))]
     /// Returns the oldest live [Record] for this [EntryHash] by getting the
     /// latest available metadata from authorities combined with this agents authored data.
+    #[cfg_attr(feature = "instrument", tracing::instrument(skip(self, options)))]
     pub async fn dht_get_entry(
         &self,
         entry_hash: EntryHash,
         options: GetOptions,
     ) -> CascadeResult<Option<Record>> {
         let query: GetLiveEntryQuery = self.construct_query_with_data_access(entry_hash.clone());
 
-        if let GetStrategy::Local = options.strategy {
-            // Only return what is in the database.
-            return self.cascading(query.clone()).await;
-        }
+        self.get_with_query(query, entry_hash.into(), options).await
+    }
 
-        // If we are not in the process of authoring this hash or its
-        // authority we need a network call.
-        let authoring = self.am_i_authoring(&entry_hash.clone().into())?;
-        let authority = self.am_i_an_authority(entry_hash.clone().into()).await?;
-        if !(authoring || authority) {
-            self.fetch_record(entry_hash.into(), options.into()).await?;
+    async fn get_with_query<Q, O>(
+        &self,
+        query: Q,
+        get_target: AnyDhtHash,
+        options: GetOptions,
+    ) -> CascadeResult<Q::Output>
+    where
+        Q: Query<Item = Judged<SignedActionHashed>, Output = Option<O>> + Send + 'static,
+        O: Send + 'static,
+    {
+        // Try to get the record from our databases first.
+        if let Some(record) = self.cascading(query.clone()).await? {
+            return Ok(Some(record));
         }
 
-        // Check if we have the data now after the network call.
-        self.cascading(query).await
+        // If we are allowed to get the data from the network then try to retrieve the missing data.
+        if options.strategy == GetStrategy::Network {
+            // If we are not in the process of authoring this hash or its
+            // authority we need a network call.
+            let authoring = self.am_i_authoring(&get_target)?;
+            let authority = self.am_i_an_authority(get_target.clone().into()).await?;
+            if !(authoring || authority) {
+                // Fetch the data if there is anyone to fetch it from.
+                match self.fetch_record(get_target, options.into()).await {
+                    Ok(_)
+                    | Err(CascadeError::NetworkError(HolochainP2pError::NoPeersForLocation(
```

Comment:
This is the only bit I'm not completely sure about. Previously this would have resulted in the get request being forwarded to us and returning no results. So it's the same behavior. Whether that's reasonable is another matter. If I don't filter this error out to maintain behavior then a handful of tests that expect to be able to do a "local" get and get a "None" response start failing. Those tests could be tweaked to actually do a local get rather than using "GetOptions::default" which permits network access.

I chose to put this up like this both because it fixes tests and because I think it's probably what I'd want as an app developer. I can't do anything about other peers being offline and I'd rather have a "this is missing" response than a network error.

Reply:
I'm happy with this. If there's any question I feel like we should err on the side of it already worked this way for the first integration. We can always refactor later.